src/ncp/ncp_base.cpp

```diff
@@ -0,0 +1,1572 @@
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ *    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ *    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+ *    DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ *    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ *    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a Spinel interface to the OpenThread stack.
+ */
+
+#include <assert.h>
+#include <common/code_utils.hpp>
+#include <ncp/ncp.hpp>
+#include <openthread.h>
+#include <openthread-config.h>
+#include <stdarg.h>
+#include <platform/radio.h>
+
+namespace Thread {
+
+extern ThreadNetif *sThreadNetif;
+
+static spinel_status_t ThreadErrorToSpinelStatus(ThreadError error)
+{
+    spinel_status_t ret;
+
+    switch (error)
+    {
+    case kThreadError_None:
+        ret = SPINEL_STATUS_OK;
+        break;
+
+    case kThreadError_Failed:
+        ret = SPINEL_STATUS_FAILURE;
+        break;
+
+    case kThreadError_Drop:
+        ret = SPINEL_STATUS_DROPPED;
+        break;
+
+    case kThreadError_NoBufs:
+        ret = SPINEL_STATUS_NOMEM;
+        break;
+
+    case kThreadError_Busy:
+        ret = SPINEL_STATUS_BUSY;
+        break;
+
+    case kThreadError_Parse:
+        ret = SPINEL_STATUS_PARSE_ERROR;
+        break;
+
+    case kThreadError_InvalidArgs:
+        ret = SPINEL_STATUS_INVALID_ARGUMENT;
+        break;
+
+    case kThreadError_NotImplemented:
+        ret = SPINEL_STATUS_UNIMPLEMENTED;
+        break;
+
+    case kThreadError_InvalidState:
+        ret = SPINEL_STATUS_INVALID_STATE;
+        break;
+
+    default:
+        ret = SPINEL_STATUS_FAILURE;
+        break;
+    }
+
+    return ret;
+}
+
+NcpBase::NcpBase():
```

Comment:
One option is to create a `NcpFramer` class with a constructor that would take by reference the `Hdlc` or `Flen` object.  This would also require creating an abstract class `Framer` for the encoder/decoder and have `Hdlc` and `Flen` inherit `Framer`.  Basically, the C++ equivalent of passing a struct of function pointers.